File: 07.compiler/code-writer.go
package main

import (
	"fmt"
	"strings"
)

func NewCodeWriter() *CodeWriter {
	return &CodeWriter{
		labelIfCount:   0,
		labelWhileCount: 0,
	}
}

func (cw *CodeWriter) WriteFunction(declaration []Tokenized, classTable []Table, parameters [][]Tokenized, className string) string {
	functionType := declaration[0].Value
	paramsLength := 0

	for _, param := range parameters {
		commaIndex := findIndex(param, ",")
		if commaIndex != -1 {
			vars := filterTokens(param[2:len(param)-1], ",")
			paramsLength += len(vars)
		} else {
			paramsLength++
		}
	}

	if functionType == "constructor" {
		functionDec := fmt.Sprintf("function %s.%s %d", declaration[1].Value, declaration[2].Value, paramsLength)
		pushConstant := fmt.Sprintf("push constant %d", len(classTable))
		callMemory := "call Memory.alloc 1"
		popPointer := "pop pointer 0"

		return fmt.Sprintf("%s\n%s\n%s\n%s", functionDec, pushConstant, callMemory, popPointer)
	} else if functionType == "method" {
		functionDec := fmt.Sprintf("function %s.%s %d", className, declaration[2].Value, paramsLength)
		argument := "push argument 0"
		popPointer := "pop pointer 0"

		return fmt.Sprintf("%s\n%s\n%s", functionDec, argument, popPointer)
	} else {
		functionDec := fmt.Sprintf("function %s.%s %d", className, declaration[2].Value, paramsLength)
		return functionDec
	}
}

func (cw *CodeWriter) WriteCall(name string, args int) string {
	return fmt.Sprintf("call %s %d", name, args)
}

func (cw *CodeWriter) WritePush(pushType, value string) string {
	return fmt.Sprintf("push %s %s", pushType, value)
}

func (cw *CodeWriter) WritePop(popType, value string) string {
	return fmt.Sprintf("pop %s %s", popType, value)
}

func (cw *CodeWriter) WriteIfLabels() map[string]string {
	labels := map[string]string{
		"if":      fmt.Sprintf("if-goto IF_TRUE%d", cw.labelIfCount),
		"goto":    fmt.Sprintf("goto IF_FALSE%d", cw.labelIfCount),
		"labelIf": fmt.Sprintf("label IF_TRUE%d", cw.labelIfCount),
		"labelElse": fmt.Sprintf("label IF_FALSE%d", cw.labelIfCount),
		"gotoEnd": fmt.Sprintf("goto IF_END%d", cw.labelIfCount),
		"labelEnd": fmt.Sprintf("label IF_END%d", cw.labelIfCount),
	}
	cw.labelIfCount++
	return labels
}

func (cw *CodeWriter) WriteWhileLabels() map[string]string {
	labels := map[string]string{
		"labelStart": fmt.Sprintf("label WHILE_EXP%d", cw.labelWhileCount),
		"if":         fmt.Sprintf("if-goto WHILE_END%d", cw.labelWhileCount),
		"goto":       fmt.Sprintf("goto WHILE_EXP%d", cw.labelWhileCount),
		"labelEnd":   fmt.Sprintf("label WHILE_END%d", cw.labelWhileCount),
	}
	cw.labelWhileCount++
	return labels
}

func (cw *CodeWriter) TransformKind(kind string) string {
	switch kind {
	case "field":
		return "this"
	case "var":
		return "local"
	default:
		return kind
	}
}

func (cw *CodeWriter) WriteString(value string) string {
	var builder strings.Builder
	builder.WriteString(fmt.Sprintf("push constant %d\n", len(value)))
	builder.WriteString("call String.new 1\n")

	for _, char := range value {
		builder.WriteString(fmt.Sprintf("push constant %d\n", char))
		builder.WriteString("call String.appendChar 2\n")
	}

	return builder.String()
}

File: draft/compiler/code-writer.ts
import { Table, Tokenized } from './types';

export class CodeWriter {
    labelIfCount: number;
    labelWhileCount: number;
    constructor() {
        this.labelIfCount = 0;
        this.labelWhileCount = 0;
    }

    public writeFunction(
        declaration: Tokenized[],
        classTable: Table[],
        subroutineTable: Table[],
        parameters: Tokenized[][],
        className: string
    ): string {
        const functionType = declaration[0].value;
        let paramsLength = 0;
        parameters.forEach((param) => {
            const commaIndex = param.findIndex((token) => token.value === ',');
            if (commaIndex !== -1) {
                const vars = param
                    .slice(2, param.length - 1)
                    .filter((token) => token.value !== ',');
                vars.forEach(() => {
                    paramsLength++;
                });
            } else {
                paramsLength++;
            }
        });
        if (functionType === 'constructor') {
            const functionDec = `function ${declaration[1].value}.${declaration[2].value} ${paramsLength}`;
            const pushConstant = `push constant ${classTable.length}`;
            const callMemory = `call Memory.alloc 1`;
            const popPointer = `pop pointer 0`;

            return `${functionDec}\n${pushConstant}\n${callMemory}\n${popPointer}`;
        } else if (functionType === 'method') {
            const functionDec = `function ${className}.${declaration[2].value} ${paramsLength}`;
            const argument = 'push argument 0';
            const popPointer = 'pop pointer 0';

            return `${functionDec}\n${argument}\n${popPointer}`;
        } else {
            const functionDec = `function ${className}.${declaration[2].value} ${paramsLength}`;
            return functionDec;
        }
    }

    writeCall(name: string, args: number): string {
        return `call ${name} ${args}`;
    }

    public writePush(type: string, value: string): string {
        return `push ${type} ${value}`;
    }

    public writePop(type: string, value: string): string {
        return `pop ${type} ${value}`;
    }

    public writeIfLabels(): {
        if: string;
        goto: string;
        labelIf: string;
        labelElse: string;
        gotoEnd: string;
        labelEnd: string;
    } {
        const labels = {
            if: `if-goto IF_TRUE${this.labelIfCount}`,
            goto: `goto IF_FALSE${this.labelIfCount}`,
            labelIf: `label IF_TRUE${this.labelIfCount}`,
            labelElse: `label IF_FALSE${this.labelIfCount}`,
            gotoEnd: `goto IF_END${this.labelIfCount}`,
            labelEnd: `label IF_END${this.labelIfCount}`,
        };
        this.labelIfCount++;

        return labels;
    }

    public writeWhileLabels(): {
        labelStart: string;
        if: string;
        goto: string;
        labelEnd: string;
    } {
        const labels = {
            labelStart: `label WHILE_EXP${this.labelWhileCount}`,
            if: `if-goto WHILE_END${this.labelWhileCount}`,
            goto: `goto WHILE_EXP${this.labelWhileCount}`,
            labelEnd: `label WHILE_END${this.labelWhileCount}`,
        };
        this.labelWhileCount++;

        return labels;
    }

    public transformKind(kind: string): string {
        switch (kind) {
            case 'field':
                return 'this';
            case 'var':
                return 'local';
            default:
                return kind;
        }
    }

    public writeString(value: string): string {
        return (
            `push constant ${value.length}\n` +
            `call String.new 1\n` +
            value
                .split('')
                .map(
                    (char) =>
                        `push constant ${char.charCodeAt(0)}\n` +
                        'call String.appendChar 2\n'
                )
                .join('')
        );
    }
}
